File: redwood-yoga-java/src/main/java/io/github/orioncraftmc/meditate/internal/YGConfig.java
package io.github.orioncraftmc.meditate.internal;

import static io.github.orioncraftmc.meditate.internal.GlobalMembers.YGNodeClone;
import io.github.orioncraftmc.meditate.internal.enums.YGExperiment;
import io.github.orioncraftmc.meditate.internal.enums.YGLogLevel;
import io.github.orioncraftmc.meditate.internal.interfaces.YGCloneNodeFunc;
import io.github.orioncraftmc.meditate.internal.interfaces.YGLogger;
import java.util.ArrayList;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class YGConfig implements Cloneable //Type originates from: YGConfig.h
{
    private final @NotNull logger_Struct logger_struct = new logger_Struct();
    public boolean useWebDefaults = false;
    public boolean useLegacyStretchBehaviour = false;
    public boolean shouldDiffLayoutWithoutLegacyStretchBehaviour = false;
    public boolean printTree = false;
    public float pointScaleFactor = 1.0f;
    public final @NotNull ArrayList<Boolean> experimentalFeatures = new ArrayList<>();
    public @Nullable Object context = null;
    private cloneNodeCallback_Struct cloneNodeCallback_struct = new cloneNodeCallback_Struct();
    private boolean cloneNodeUsesContext_;
    private boolean loggerUsesContext_;

    public YGConfig(YGLogger logger) //Method definition originates from: YGConfig.cpp
    {
        this.cloneNodeCallback_struct = null;
        logger_struct.noContext = logger;
        loggerUsesContext_ = false;
        for (int i = 0; i < YGExperiment.values().length; i++) {
            experimentalFeatures.add(false);
        }
    }

    @Override
    public @NotNull YGConfig clone() {
        try {
            return (YGConfig) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new RuntimeException(e);
        }
    }

    public final void log(YGConfig config, YGNode node, YGLogLevel logLevel, Object logContext, String format, Object... args) //Method definition originates from: YGConfig.cpp
    {
        if (loggerUsesContext_) {
            logger_struct.withContext.invoke(config, node, logLevel, logContext, format, args);
        } else {
            logger_struct.noContext.invoke(config, node, logLevel, format, args);
        }
    }

    public final void setLogger(YGLogger logger) {
        logger_struct.noContext = logger;
        loggerUsesContext_ = false;
    }

    public final void setLogger(LogWithContextFn logger) {
        logger_struct.withContext = logger;
        loggerUsesContext_ = true;
    }

    public final void setLogger() {
        logger_struct.noContext = null;
        loggerUsesContext_ = false;
    }

    public final @NotNull YGNode cloneNode(YGNode node, YGNode owner, int childIndex, Object cloneContext) //Method definition originates from: YGConfig.cpp
    {
        @Nullable YGNode clone = null;
        if (cloneNodeCallback_struct.noContext != null) {
            clone = cloneNodeUsesContext_ ? cloneNodeCallback_struct.withContext.invoke(node, owner, childIndex,
                    cloneContext) : cloneNodeCallback_struct.noContext.invoke(node, owner, childIndex);
        }
        if (clone == null) {
            clone = YGNodeClone(node);
        }
        return clone;
    }

    public final void setCloneNodeCallback(YGCloneNodeFunc cloneNode) {
        cloneNodeCallback_struct.noContext = cloneNode;
        cloneNodeUsesContext_ = false;
    }

    public final void setCloneNodeCallback(CloneWithContextFn cloneNode) {
        cloneNodeCallback_struct.withContext = cloneNode;
        cloneNodeUsesContext_ = true;
    }

    public final void setCloneNodeCallback() {
        cloneNodeCallback_struct.noContext = null;
        cloneNodeUsesContext_ = false;
    }

    @FunctionalInterface
    public interface LogWithContextFn {
        int invoke(YGConfig config, YGNode node, YGLogLevel level, Object context, String format, Object... args);
    }

    @FunctionalInterface
    public interface CloneWithContextFn {
        @NotNull YGNode invoke(YGNode node, YGNode owner, int childIndex, Object cloneContext);
    }

    private static class cloneNodeCallback_Struct {

        CloneWithContextFn withContext;
        @Nullable YGCloneNodeFunc noContext;

    }

    private static class logger_Struct {

        LogWithContextFn withContext;
        @Nullable YGLogger noContext;

    }
}

File: redwood-yoga/src/commonMain/kotlin/app/cash/redwood/yoga/internal/YGConfig.kt
package app.cash.redwood.yoga.internal

import app.cash.redwood.yoga.internal.enums.YGExperiment
import app.cash.redwood.yoga.internal.enums.YGLogLevel
import app.cash.redwood.yoga.internal.interfaces.YGCloneNodeFunc
import app.cash.redwood.yoga.internal.interfaces.YGLogger

class YGConfig(logger: YGLogger?) //Type originates from: YGConfig.h
{
    private val logger_struct = logger_Struct()
    var useWebDefaults = false
    var useLegacyStretchBehaviour = false
    var shouldDiffLayoutWithoutLegacyStretchBehaviour = false
    var printTree = false
    var pointScaleFactor = 1.0f
    val experimentalFeatures = ArrayList<Boolean>()
    var context: Any? = null
    private var cloneNodeCallback_struct: cloneNodeCallback_Struct? = cloneNodeCallback_Struct()
    private var cloneNodeUsesContext_ = false
    private var loggerUsesContext_: Boolean

    init  //Method definition originates from: YGConfig.cpp
    {
        cloneNodeCallback_struct = null
        logger_struct.noContext = logger
        loggerUsesContext_ = false
        for (i in YGExperiment.values().indices) {
            experimentalFeatures.add(false)
        }
    }

    fun log(
      config: YGConfig?,
      node: YGNode?,
      logLevel: YGLogLevel,
      logContext: Any?,
      format: String,
      vararg args: Any?
    ) //Method definition originates from: YGConfig.cpp
    {
        if (loggerUsesContext_) {
            logger_struct.withContext!!.invoke(config, node, logLevel, logContext, format, *args)
        } else {
            logger_struct.noContext!!.invoke(config, node, logLevel, format, *args)
        }
    }

    fun setLogger(logger: YGLogger?) {
        logger_struct.noContext = logger
        loggerUsesContext_ = false
    }

    fun setLogger(logger: LogWithContextFn?) {
        logger_struct.withContext = logger
        loggerUsesContext_ = true
    }

    fun setLogger() {
        logger_struct.noContext = null
        loggerUsesContext_ = false
    }

    fun cloneNode(
      node: YGNode,
      owner: YGNode?,
      childIndex: Int,
      cloneContext: Any?
    ): YGNode //Method definition originates from: YGConfig.cpp
    {
        var clone: YGNode? = null
        if (cloneNodeCallback_struct!!.noContext != null) {
            clone = if (cloneNodeUsesContext_) cloneNodeCallback_struct!!.withContext!!.invoke(
                node, owner, childIndex,
                cloneContext
            ) else cloneNodeCallback_struct!!.noContext!!.invoke(node, owner, childIndex)
        }
        if (clone == null) {
            clone = GlobalMembers.YGNodeClone(node)
        }
        return clone
    }

    fun setCloneNodeCallback(cloneNode: YGCloneNodeFunc?) {
        cloneNodeCallback_struct!!.noContext = cloneNode
        cloneNodeUsesContext_ = false
    }

    fun setCloneNodeCallback(cloneNode: CloneWithContextFn?) {
        cloneNodeCallback_struct!!.withContext = cloneNode
        cloneNodeUsesContext_ = true
    }

    fun setCloneNodeCallback() {
        cloneNodeCallback_struct!!.noContext = null
        cloneNodeUsesContext_ = false
    }

    fun interface LogWithContextFn {
        operator fun invoke(
          config: YGConfig?,
          node: YGNode?,
          level: YGLogLevel?,
          context: Any?,
          format: String?,
          vararg args: Any?
        ): Int
    }

    fun interface CloneWithContextFn {
        operator fun invoke(
          node: YGNode?,
          owner: YGNode?,
          childIndex: Int,
          cloneContext: Any?
        ): YGNode
    }

    internal class cloneNodeCallback_Struct {
        var withContext: CloneWithContextFn? = null
        var noContext: YGCloneNodeFunc? = null
    }

    internal class logger_Struct {
        var withContext: LogWithContextFn? = null
        var noContext: YGLogger? = null
    }
}
